include/podio/detail/RelationIOHelpers.h

```diff
@@ -0,0 +1,178 @@
+#ifndef PODIO_DETAIL_RELATIONIOHELPERS_H
+#define PODIO_DETAIL_RELATIONIOHELPERS_H
+
+#include "podio/utilities/TypeHelpers.h"
+#include <podio/CollectionBase.h>
+
+#include <tuple>
+#include <vector>
+
+namespace podio::detail {
+
+/// Function template for handling interface types in OneToMultiRelations
+///
+/// Effectively this function checks whether the passed collection can be
+/// dynamically cast to the collection type of the concrete type and if that is
+/// true uses it to construct an interface type and add it to relElements. The
+/// function on its own doesn't do anything too meaningful, it is meant to be
+/// used in a call to std::apply that goes over all the interfaced types of an
+/// interface type.
+///
+/// @tparam T The concrete type inside the interface that should be checked.
+///           This effectively is mainly used for tag-dispatch and overload
+///           selection in this context
+/// @tparam InterfaceType The interface type (that can be used to interface T)
+///
+/// @param relElements The vector to which the interface objects should be added
+/// @param coll The collection that holds the actual element
+/// @param id The ObjectID of the element that we are currently looking for
+template <typename T, typename InterfaceType>
+void tryAddTo(T, std::vector<InterfaceType>& relElements, const podio::CollectionBase* coll, const podio::ObjectID id) {
+  if (auto typedColl = dynamic_cast<const typename T::collection_type*>(coll)) {
+    const T tmp = (*typedColl)[id.index];
+    relElements.emplace_back(tmp);
+  }
+}
+
+/// Helper function for handling interface type relations in OneToManyRelations
+///
+/// This function tries all types that are interfaced by the InterfaceType and
+/// adds the one that matches to the relations. The main work happens in
+/// tryAddTo, this simply wraps everything in a std::apply over all
+/// interfaced_types.
+///
+/// @tparam InterfaceType The interface type of the Relation
+///
+/// @param relElements The vector to which the interface objects should be added
+/// @param coll The collection that holds the actual element
+/// @param id The ObjectID of the element that we are currently looking for
+template <typename InterfaceType>
+void addInterfaceToMultiRelation(std::vector<InterfaceType>& relElements, const podio::CollectionBase* coll,
+                                 const podio::ObjectID id) {
+  std::apply([&relElements, &coll, &id](auto... t) { (tryAddTo(t, relElements, coll, id), ...); },
```

Comment:
```suggestion
  std::apply([&](auto... t) { (tryAddTo(t, relElements, coll, id), ...); },
```
since we capture everything in any case